### ISSUE TITLE
ImGui windows becoming a mask for SDL_RenderDrawLine() Calls

### ISSUE THREAD

**Issue by sufiblues** [ocornut/imgui#4582]
Hello, I am having an issue with my gui set up. I am trying to creatin a simple window but dearimgui window is a mask for the actual game scene. 
I am using SDL 2.0.17
SDL2 + SDL_Renderer backend 
Here is what is looks like
![Screenshot from 2021-09-24 17-58-34](https://user-images.githubusercontent.com/70812929/134747829-7731a576-a42d-46c2-96af-31fa9d200949.png)
![Screenshot from 2021-09-24 17-58-04](https://user-images.githubusercontent.com/70812929/134747835-18d93ac2-ebb5-477b-86cf-61b489fc7015.png)

![Screenshot from 2021-09-24 17-58-00](https://user-images.githubusercontent.com/70812929/134747838-1d3b72d8-8002-46f2-9476-285985ec7c7a.png)

Here is the code that I am using to generate the scene
```

void render(){

  // start the imgui frame
  ImGui_ImplSDLRenderer_NewFrame();
  ImGui_ImplSDL2_NewFrame(Window);
  ImGui::NewFrame();

  ImGui::Begin("Scene");
  ImGui::Text("Temp");
  ImGui::End();
  ImGui::EndFrame();

  SDL_SetRenderDrawColor(Renderer, 0, 0, 0, 0);
  SDL_RenderClear(Renderer);

  SDL_Rect o = {40,40,100,40};
  SDL_SetRenderDrawColor(Renderer, 100, 32, 13, 255);
  SDL_RenderFillRect(Renderer, &o);

  ImGui::Render();
  ImGui_ImplSDLRenderer_RenderDrawData(ImGui::GetDrawData());

  RenderGameState(&game); //A bunch of SDL_RenderDrawLines() 

  //SDL_GL_SwapWindow(Window);
  SDL_RenderPresent(Renderer);
}
```

**Comment by PathogenDavid** [comment]
Does your app work without rendering the UI?

A few things stand out to me in your code:

> ```c
> SDL_Rect o = {40,40,100,40};
> SDL_SetRenderDrawColor(Renderer, 100, 32, 13, 255);
> SDL_RenderFillRect(Renderer, &o);
> ```

This rectangle doesn't appear to be visible either. (Hence my question about it working without the UI.)

> ```c
> SDL_SetRenderDrawColor(Renderer, 0, 0, 0, 0);
> SDL_RenderClear(Renderer);
> `

You're clearing the alpha of the render target to 0 instead of 255. You probably don't actually want to do that.

> ```c
> ImGui::Render();
> ImGui_ImplSDLRenderer_RenderDrawData(ImGui::GetDrawData());
> 
> RenderGameState(&game); //A bunch of SDL_RenderDrawLines() 
> ```

You're rendering your game after the UI. Even if it was working correctly, you'd be rendering on top of it.

------------

Also keep in mind that SDL_Renderer is a stateful API. If you change state (such as the blend mode) in `RenderGameState` it will carry over to subsequent frames (and potentially cause rendering issues if it wasn't what you expected.)

**Comment by ocornut** [comment]
That new sdl renderer doesn’t restore clip rect after rendering I think that’s the reason.

**Comment by PathogenDavid** [comment]
Oh yeah that'd definitely do it.

@sufiblues You should be able to add a call to `SDL_RenderSetClipRect(Renderer, NULL);` after `ImGui_ImplSDLRenderer_RenderDrawData` as a workaround until the backend is fixed.

**Comment by sufiblues** [comment]
Thanks, changing the order of the rendering calls as you suggested @PathogenDavid  and calling SDL_RenderClipRect(Renderer, NULL); after rendering the gui fixed the issue. Super excited about integrating imgui with my game.

**Comment by ocornut** [comment]
Now pushed a fix to backup/restore that state e443ea1
